MAM: rough draft

mcneb10 opened the pull request:
Here is my take on the MAM implementation so far

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+        fl.append(end);
+    }
+
+    // TODO: ID filters?
```
most likely those will be required. 
imagine a situation you chatted with using Conversations a lot and everything was written to MAM.
Then you switched to Psi (which was offline till that moment) and loaded just the most recent history from server.
Then your buddy sends you a message with a quote which has message ID referring your message. Psi doesn't have this message in local database and it doesn't know when it was sent, so it needs to fetch it from MAM by ID and likely some messages before and after it to show in the log.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+    return x;
+}
+
+// TODO: use iris rsm implementation
```
yup

mcneb10 replied:
i'm going to have to modify it. it currently isn't flexible enough for this use case

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+    } else {
+        if(!iqVerify(x, QString(), id())) return false;
+
+        if(x.elementsByTagName("start").at(0).isNull() || x.elementsByTagName("end").at(0).isNull() ) return false;
```
I understand any string literal by default will be interpreted at latin1 in Qt default. but it would be safer to use `QLatin1String` in all such cases. Unless string literal overload exists.

mcneb10 replied:
When is this necesary? Looking through the iris code, most of the code doesn't use `QLatin1String`

Ri0n replied:
most of the code is quite legacy.
The thing is that methods accepting QString will treat string literal as UTF-8.  QString keeps internal data as UTF-16. 
Converting data from UTF-8 to UTF-16 is in general slower than from latin1 when we don't need to check high bits. In case of string comparisons it's even more preferable to use QLatin1String coz there are overloads of comparison function which bypasses string re-encoding completely.
Yes it may be considered negligible in some particular case , but shall we ignore any optimization completely?

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.h:
```diff
@@ -0,0 +1,66 @@
+#ifndef XMPP_MAM_TASK_H
+#define XMPP_MAM_TASK_H
+
+#include <QDateTime>
```
we prefer to use includes in next order

- current dir
- anyhow local for the project
- system libs
- standard stuff like STL

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.h:
```diff
@@ -0,0 +1,66 @@
+#include <QStringList>
+#include <qdom.h>
+
+#include "xmpp/jid/jid.h"
```
I see you already tried to cleanup headers. but not all of them.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.h:
```diff
@@ -0,0 +1,66 @@
+
+    const QList<QDomElement> &archive() const;
+
+    void get(const Jid &j, const QDateTime &from = QDateTime(), const QDateTime &to = QDateTime(),
```
maybe you should review it. If you add more filters it will turn out into too many parameters. not that convenient.
In case it will make more sense to put common data sets in structures and likely use overloads and different task types for different MAM requests.

mcneb10 replied:
i was thinking that if i put this behind a new `MAMManager` class i would keep this here but abstract it away from the developer. It's nice to have a way to initialize everything in `Private` in one go

Ri0n replied:
till the moment the parameters become mutually exclusive. something to think about later

mcneb10 replied:
i made seperate get methods for time and ID filtering. The API is a little questionable, let me know if you want me to change it

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.h:
```diff
@@ -0,0 +1,66 @@
+    bool take(const QDomElement &);
+
+private:
+    void  getPage();
+    void  getArchiveMetadata();
+    XData makeMAMFilter();
```
something to move to `Private::`

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+        if (x.attribute("id") == d->lastID || d->messagesFetched >= d->mamMaxMessages) {
+            setSuccess();
+        } else if (d->messagesFetched % d->mamPageSize == 0) { 
+            getPage();
```
what if `d->messagesFetched == 0` ?

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+    bool      allowMUCArchives;
+    int mamPageSize; // TODO: this is the max page size for MAM request. Should be made into a config option
+    int       mamMaxMessages; // maximum mam pages total, also should be config
+    bool flipPages;
```
nit: it's preferable to keep same data types close to each other. 
it will spare you a few bytes. 
In this case it's not really important because we don't call it often in parallel.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+#include "xmpp_mamtask.h"
```
license header is missing

mcneb10 commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+    bool backwards;
+
+    QList<QDomElement> archive;
+    int                messagesFetched;
```
assuming we're going to be able to use the `MAMTask` class to fetch multiple MAM archives, it's important the get method contains the initialization to reset the variables

Ri0n replied:
Usually we don't reuse tasks to send "get" requessts.
But we have tasks to receive "get" requests from server and those are reused. 
breaking consistency is not that good idea wrt future maintainability of the code.

mcneb10 commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,178 @@
+XData MAMTask::makeMAMFilter() {
+    XData::FieldList fl;
+
+    XData::Field type;
+    type.setType(XData::Field::Field_Hidden);
+    type.setVar("FORM_TYPE");
+    type.setValue(QStringList("urn:xmpp:mam:2"));
+    fl.append(type);
```
i looked through `xmpp_xdata.cpp` and i couldn't find the setter for this property. It also isn't ever actually used anywhere it seems. The "FORM_TYPE" needs to be in the XML so the MAM request is valid.

Ri0n replied:
try github search. it works quite well https://github.com/search?q=repo%3Apsi-im%2Firis%20setRegistrarType&type=code

Ri0n replied:
if it feels useful, I just merged master into the original mam branch.

mcneb10 replied:
thanks, i'll figure out where it needs to go when the worker code is done

Neustradamus commented:
Good to see the @mcneb10 PR :)

Commit:
move here from the mam branch, fix filtering, semantics, logic, etc

Commit:
Merge remote-tracking branch 'refs/remotes/upstream/master'

mcneb10 commented:
I just made a [https://github.com/psi-im/iris/pull/74](new PR), will close this one once all conversations are resolved

Commit:
add id filter

Commit:
Merge branch 'master' into mam
merge updated mam branch with mam code

Commit:
actually add the id filtering logic

Commit:
some code cleanup

mcneb10 commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,236 @@
+        if (iqVerify(x, QString(), id()))
+            return false;
+
+        // TODO: only save messages directed at the correct resource?
```
Filter by resource

mcneb10 commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,236 @@
+        } else if (d->messagesFetched % d->mamPageSize == 0) {
+            d->getPage(this);
+        }
+        // TODO: handle server not sending all MAM stuff gracefully (such as handling 0 messagesFetched)
```
Error handling

mcneb10 replied:
@Ri0n how does iris handle tasks that hang too long? If the server sends a malformed response, the task will simply stop running, and the messages that were received properly can be accessed with the `archive()` getter.

Ri0n replied:
there is 2 minutes timeout

Commit:
add better error handling

mcneb10 commented:
@Ri0n  This PR is almost ready. I do have a few questions though.
1. How will I test my code?
2. Should I write a `Manager` wrapper class around the task?

Commit:
add `xmpp_mamtask` to cmake and fix compilation errors

Ri0n commented:
wrt testing Psi part is not ready yet, but Psi can load history from local db.
So we need some code in Psi which will sync MAM with local db, It will be the first step. 
If you want to test something right in iris, I believe you need to write tests yourself. Psi has miniclient which can be ported to some integration tests in iris.

Manager sounds like a good idea, having API most suitable to be used by local db synchronizer.  Like querying messages is just the first step and eventually it will have a lot of MAM control methods.

mcneb10 commented:
alright, i'll add a `MAMManager` class that simplifies creating querying the MAM archives. As for the testing, I'll try to write a small plugin or patch or something to test the querying

Commit:
add mam manager (untested, compiles)

mcneb10 commented:
@Ri0n if it doesn't bother you, can you help me out by modifying a Psi+ plugin or something to test my code? I'm not familiar with the Psi+ codebase, and some other recent changes to Iris are preventing Psi+ from compiling.

Neustradamus commented:
@Ri0n: Really important to finish this XEP support...

Ri0n commented:
@mcneb10, that was my plan.
I changed my job a few days ago and now completely consume by it. So I can't promise any quick solution.
I have some ideas in mind though. 
1. from UI point of view endless chat scrolling must be supported
2. we need a sort of message buffer per chat. A buffer which keeps for example max 200 messages. It becomes a sort of model for chatview
3. the model should be signalled from chatview when it needs to loads more messages (more recent or more old)
4. the model should signal back its busy status
5. the model queries some sort of Psi's storage/archive manager capable to interact with both local database and iris' mam manager
6. the manager will detect which exactly time spans are missed in the local db and query iris to retrieve the remaining
7. local db itself has to be modified to reflect MAM use case and also to keep track of missed time spans (e.g. a message/event can have a state like "edge event" when it was first or last in the result set unless we query from a boundary of another set)
8. wrt iris. it would be nice to have some methods in the manager like for example `loadBeforeStanzaId(string stanzaId, int limit=100)` and `loadAfterStanzaId`. This would secure the endless scrolling more carefully than timestamps.

mcneb10 commented:
> @mcneb10: What do you think about @Ri0n comment?

I'm not available to work right now, but i did add some of the changes rion requested. I'll push them in a day or two

Commit:
add ability to get last messages from archive, fix some naming

Commit:
add ability to get messages before id

mcneb10 commented:
@Ri0n i just added the code to iris which should serve the function of what you want in point #8. The two functions `getLatestMessagesFromArchive` and `getMessagesBeforeID` i added always scroll backwards, meaning they send the messages in the order of newest to oldest. The other functions have a configurable direction (set both flipPages and backwards to false and it will get messages in oldest to newest order).

As for only fetching messages in a specific time frame, you can simply do `getArchiveByTimeRange` and set the `from` to `QDateTime()` and `to` to a specific time to get all messages up to a specific point, respecting the order set in the manager constructor.
It also applies the other way, so if you set `to` to `QDateTime()` and `from` to a specific time it will get all messages starting at the time you set until the latest message.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,84 @@
+// TODO: review the safety of these methods/object lifetimes
+void MAMManager::getFullArchive(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const bool allowMUCArchives)
+{
+    MAMTask task(d->client->rootTask());
```
you create it on stack, so it will be destroyed before it has any chance to get any response from server.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,84 @@
+MAMManager::~MAMManager() { delete d; }
+
+// TODO: review the safety of these methods/object lifetimes
+void MAMManager::getFullArchive(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const bool allowMUCArchives)
```
are you absolutely sure `archiveHandler` is a good idea?
it will force do life checks in the subscriber while Qt provides those out of the box with its pubsub mechanism.
moreover function pointers limit usage of functors. which most likely will be required in this case unless you want to provide some user data in an additional parameter.

mcneb10 replied:
no, but i need a way to send the result of the query asynchronously. Is there a preferred way to handle this sort of thing within Iris?

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,84 @@
+void MAMManager::getFullArchive(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const bool allowMUCArchives)
+{
+    MAMTask task(d->client->rootTask());
+    connect(&task, &MAMTask::finished, this, [task, archiveHandler]() { archiveHandler(task.archive()); });
```
your handlers are the same everywhere and they do nothing with the manager.
maybe it worth it to just return the task as is instead of having this archiveHandler. Yeap it feels leaky but it will simplify the handling itself.
Later will can try to inject local DB interfaces here. So mam manager would directly communicate with local db, or otherwise go with signals approach when mammanager signals which part of archive became available, so local db manager may try to sync it

mcneb10 replied:
I can put that in for now

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,84 @@
+    task.get(j, QString(), QString(), allowMUCArchives, d->mamPageSize, d->mamMaxMessages, d->flipPages, d->backwards);
+}
+
+void MAMManager::getArchiveByIDRange(void (*archiveHandler)(QList<QDomElement>), const Jid &j, const QString &from_id,
```
it's  interesting in what case we can use this method.

mcneb10 replied:
this one can be used to help fetch messages from incomplete archives. for example, when the user wants to scroll up to get more messages or the user reopens the chat and needs to load the newest messages

Ri0n replied:
I understand your examples. 
but this method is unlikely to be used because you have more convenient methods below.

maybe will be useful for something. maybe not in Psi.

mcneb10 replied:
I guess I'll just keep it in there for now, I don't see a reason it needs to be removed

Commit:
make the manager just return the activated MAM task instead of handling the result

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,88 @@
+    MAMTask task(d->client->rootTask());
+
+    task.get(j, QString(), QString(), allowMUCArchives, d->mamPageSize, d->mamMaxMessages, d->flipPages, d->backwards);
+    return task;
```
QObject is technically a not copy-able thing. This means something will be broken when we get result from this function.
instead prefer `auto task  = new MAMTask(d->client->rootTask());`
You can also make it self-delete by calling the corresponding method, or let the caller do this.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,255 @@
+    iq.appendChild(metadata);
+    iq.appendChild(makeMAMFilter().toXml(t->doc()));
+
+    t->send(iq);
```
In general the approach with tasks is to remember parameters first on call to some public method, and make and send DOM element in the re-implemented `void onGo();`
eventually the task will be started by the called with `go()` method.

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,255 @@
+            d->getPage(this);
+        }
+    } else {
+        if (!iqVerify(x, QString(), id()) || x.elementsByTagName(QLatin1String("metadata")).isEmpty())
```
I think it worth moving metadata part to a separate task and initiate it somewhere on client login from manager. 
If there are multiple simultaneous archive queries while metadata is not completed yet, those can be queued somewhere in the manager (for example `onGo` can put in a queue instead of `send()`)

mcneb10 replied:
I'd say it should be kept this way until we can prove it works

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,255 @@
+
+        // Return if the archive is empty
+        if (!x.elementsByTagName(QLatin1String("metadata")).at(0).hasChildNodes()) {
+            setError(1, QLatin1String("Archive is empty"));
```
why is it considered to be an error?

Ri0n commented on src/xmpp/xmpp-im/xmpp_mamtask.cpp:
```diff
@@ -0,0 +1,255 @@
+    return x;
+}
+
+void MAMTask::Private::getPage(MAMTask *t)
```
better pass the task once to the Private's constructor and remember there. You `q` member for that, which is sort of standard

mcneb10 commented:
I'm going to be pretty busy for a while. I won't fix these issues for a few weeks

mcneb10 commented:
Im going to see if i can finish this up

Commit:
fix archive behavior and return value for mam manager

mcneb10 commented:
I will not be able to get anywhere near finishing this code until there is a mechanism added somewhere in Psi+ to call my code

Neustradamus commented:
@Ri0n: Can you check?

mcneb10 commented:
It's important to know that the latest version of iris does not compile with Psi+ due to some changes to the carbons code

Ri0n commented on src/xmpp/xmpp-im/xmpp_mammanager.cpp:
```diff
@@ -0,0 +1,88 @@
+{
+    MAMTask task = new MAMTask(d->client->rootTask());
+
+    task.get(j, from, to, allowMUCArchives, d->mamPageSize, d->mamMaxMessages, d->flipPages, d->backwards);
```
you can try to build iris separately. At least you'll fix compilation errors.

mcneb10 replied:
i have, and it does compile nicely. is there a way to link it with Psi+ without building it through `psi-plus-snapshots`? The psi+ source code in there seems to be reliant on an alternate implementation of carbons that no longer works within iris

Ri0n replied:
ah. I just got it. Implicit constructor call made it compile.

basically what happens...
```
MAMTask task = new MAMTask(d->client->rootTask());
```
this line creates 2 tasks. first one with `new`, and the second one with a implicit constructor call when we assign it to the `task` variable.
You can check it marking the corresponding constructor as `explicit`
```
explicit MAMTask(Task *);
```

Ri0n commented:
fixed in mam branch

mcneb10 commented:
Thank you for the fix! This is very exciting.

Neustradamus commented:
@mcneb10: Good job, an important part has been done!

Thanks @Ri0n for merging, but why not in master?

Can wait next parts ^^

mcneb10 commented:
I'd say it needs testing first

Neustradamus commented:
A new XEP-0313 version has been released - 1.1.2 (2024-09-18):
- https://xmpp.org/extensions/xep-0313.html
- https://www5.aptest.com/standards/htmldiff/htmldiff.pl?oldfile=https://xmpp.org/extensions/attic/xep-0313-1.1.1.html&newfile=https://xmpp.org/extensions/attic/xep-0313-1.1.2.html

It is good?
